plugins/com.google.cloud.tools.eclipse.test.util/src/com/google/cloud/tools/eclipse/test/util/project/ProjectUtils.java

```diff
@@ -183,49 +186,67 @@ public void run() {
 
   /** Wait for any spawned jobs and builds to complete (e.g., validation jobs). */
   public static void waitForProjects(Runnable delayTactic, IProject... projects) {
-    IJobManager jobManager = Job.getJobManager();
+    Stopwatch timer = Stopwatch.createStarted();
     try {
-      Set<Job> jobs = new LinkedHashSet<>();
-      List<String> allBuildErrors;
+      Collection<Job> jobs = Collections.emptyList();
+      Set<String> previousBuildErrors = Collections.emptySet();
+      boolean buildErrorsChanging;
       do {
+        // wait a little bit to give the builders a chance
         delayTactic.run();
+        
+        // wait for any previously-identified build jobs
         for (Job job : jobs) {
           job.join();
         }
-        jobs.clear();
-
-        Collections.addAll(jobs, jobManager.find(ResourcesPlugin.FAMILY_MANUAL_BUILD));
-        Collections.addAll(jobs, jobManager.find(ResourcesPlugin.FAMILY_AUTO_BUILD));
-        // J2EEElementChangedListener.PROJECT_COMPONENT_UPDATE_JOB_FAMILY
-        Collections.addAll(jobs, jobManager.find("org.eclipse.jst.j2ee.refactor.component"));
-        // ServerPlugin.SHUTDOWN_JOB_FAMILY
-        Collections.addAll(jobs, jobManager.find("org.eclipse.wst.server.core.family"));
-        Collections.addAll(jobs, jobManager.find("org.eclipse.wst.server.ui.family"));
-        Collections.addAll(jobs, jobManager.find(ValidationBuilder.FAMILY_VALIDATION_JOB));
-        allBuildErrors = getAllBuildErrors(projects);
-        if (DEBUG) {
+
+        // identify any pending build-related jobs
+        jobs = findPendingBuildJobs();
+
+        // track whether we've reached a fixpoint in the build errors
+        Set<String> currentBuildErrors = getAllBuildErrors(projects);
+        buildErrorsChanging = !previousBuildErrors.equals(currentBuildErrors);
+        previousBuildErrors = currentBuildErrors;
+
+        if (DEBUG || timer.elapsed(TimeUnit.SECONDS) > 10) {
```

Comment:
But we sleep, and we only log for the conditions that keep us in the loop which is important info.